server/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -204,13 +203,23 @@ public String toString() {
         });
     }
 
-    private void onPrimaryDemoted(Exception demotionFailure) {
-        String primaryFail = String.format(Locale.ROOT,
-            "primary shard [%s] was demoted while failing replica shard",
-            primary.routingEntry());
-        // we are no longer the primary, fail ourselves and start over
-        primary.failShard(primaryFail, demotionFailure);
-        finishAsFailed(new RetryOnPrimaryException(primary.routingEntry().shardId(), primaryFail, demotionFailure));
+    private void onNoLongerPrimary(Exception failure) {
+        final boolean nodeIsClosing = failure instanceof NodeClosedException ||
+            (failure instanceof TransportException && "TransportService is closed stopped can't send request".equals(failure.getMessage()));
+        final String message;
+        if (nodeIsClosing) {
+            message = String.format(Locale.ROOT, "primary node [%s] is shutting down while failing replica shard", primary.routingEntry());
```

Comment:
perhaps change to "node with primary [%s] is shutting ..."

Reply:
Pushed https://github.com/elastic/elasticsearch/pull/39584/commits/6b56fbcee0171e995a01ac7f1ba288963be44d41.